### PR TITLE
etcdctlv3: improve txn interactive mode

### PR DESCRIPTION
**Pull request by heyitsanthony**

### CODE REVIEW

File: etcdctlv3/README.md [hunk 1]
```diff
@@ -109,6 +109,61 @@ OK
 ./etcdctl range foo
 ```
 
+### TXN [options]
+
+Watch watches events stream on keys or prefixes. The watch command runs until it encounters an error or is terminated by the user.
```
Comment (xiang90) [inline etcdctlv3/README.md]: update this description?

File: etcdctlv3/README.md [hunk 2]
```diff
@@ -109,6 +109,61 @@ OK
...
+
+#### Options
+
+- hex -- print out keys and values as hex encoded string
+
+- i -- input transaction with interactive mode
+
+#### Input Format
```
Comment (xiang90) [inline etcdctlv3/README.md]: is this for interactive mode only?

File: etcdctlv3/README.md [hunk 3]
```diff
@@ -109,6 +109,61 @@ OK
...
+<THEN> ::= Op*
+<ELSE> ::= Op*
+<OP> ::= see put, get, del etcdctl command syntax
+```
+
+#### Return value
+
+Simple reply
+
+- SUCCESS if etcd processed the transaction success list, FAILURE if etcd processed the transaction failure list.
```
Comment (xiang90) [inline etcdctlv3/README.md]: is there a \r\n follows SUCCESS or FAILURE before the simple reply of requests?

File: etcdctlv3/README.md (hunk 1 above)
Comment (heyitsanthony) [inline etcdctlv3/README.md]: Oops

File: etcdctlv3/README.md [hunk 4]
```diff
@@ -109,6 +109,61 @@ OK
...
+
+TODO: probably json and binary encoded proto
+
+#### Examples
+
+``` bash
+./etcdctl txn -i
+key1 greater mod 0
+
+put key1 "overwrote-key1"
```
Comment (xiang90) [inline etcdctlv3/README.md]: for the interactive mode, probably we can print out hints? Like cmp:\n, success: \n, failure: \n?

File: etcdctlv3/command/txn_command.go [hunk 5]
```diff
@@ -18,21 +18,31 @@ import (
...
 func NewTxnCommand() *cobra.Command {
-	return &cobra.Command{
-		Use:   "txn",
+	cmd := &cobra.Command{
+		Use:   "txn [options]",
 		Short: "Txn processes all the requests in one transaction.",
 		Run:   txnCommandFunc,
 	}
+	cmd.Flags().BoolVar(&txnInteractive, "i", false, "input transaction in interactive mode")
+	cmd.Flags().BoolVar(&txnHex, "i", false, "input transaction in interactive mode")
```
Comment (xiang90) [inline etcdctlv3/command/txn_command.go]: cmd.Flags().BoolVar(&txnHex, "hex", false, "accept and print out key-value as hex encode string")

File: etcdctlv3/README.md (hunk 4 above)
Comment (xiang90) [inline etcdctlv3/README.md]: ignore this. we do have hint lines.

File: etcdctlv3/command/txn_command.go [hunk 6]
```diff
@@ -41,198 +51,178 @@ func txnCommandFunc(cmd *cobra.Command, args []string) {
...
 		return nil, fmt.Errorf("invalid txn compare request: %s", line)
 	}
-	return c, nil
+
+	return &cmp, nil
+}
+
+func printTxnResponse(resp clientv3.TxnResponse, isHex bool) {
+	if resp.Succeeded {
+		fmt.Println("SUCCESS")
```
Comment (xiang90) [inline etcdctlv3/command/txn_command.go]: we use \r\n to do the separation in other commands. should we adopt that format?

File: etcdctlv3/README.md (hunk 2 above)
Comment (heyitsanthony) [inline etcdctlv3/README.md]: yes, will make clear

**Comment by xiang90** [comment]
Code change looks good to me. I am going to give this a try to see if there is any issue and get back to you soon.

File: etcdctlv3/README.md (hunk 3 above)
Comment (heyitsanthony) [inline etcdctlv3/README.md]: I've been using `Println` which is '\n'. Are we standardizing on '\r\n'?

File: etcdctlv3/command/txn_command.go (hunk 5 above)
Comment (heyitsanthony) [inline etcdctlv3/command/txn_command.go]: oops, thanks

File: etcdctlv3/README.md (hunk 3 above)
Comment (xiang90) [inline etcdctlv3/README.md]: to do the separation, yes. but mostly, i just want to have consistency among commands. if you want \n, it is fine. But we need to change others.

File: etcdctlv3/command/txn_command.go (hunk 5 above)
Comment (xiang90) [inline etcdctlv3/command/txn_command.go]: i was wrong. this is only for printing.

File: etcdctlv3/README.md (hunk 3 above)
Comment (heyitsanthony) [inline etcdctlv3/README.md]: Yeah, '\n' is less typing / less verbose and is what Println does. I'll do a separate PR to remove teletype support.

Comment (xiang90) [inline etcdctlv3/README.md]: sure.

File: etcdctlv3/README.md [hunk 7]
```diff
@@ -109,6 +109,64 @@ OK
...
+
+#### Options
+
+- hex -- print out keys and values as hex encoded string
+
+- i -- input transaction with interactive mode
+
+#### Input Format
+Interactive mode:
+```bnf
```
Comment (xiang90) [inline etcdctlv3/README.md]: this is ebnf?

Comment (xiang90) [inline etcdctlv3/README.md]: anyway... we probably do not want this strict bnf

File: etcdctlv3/README.md [hunk 8]
```diff
@@ -109,6 +109,64 @@ OK
...
 
+### TXN [options]
+
+TXN applies multiple etcd requests as a single atomic transaction. A transaction consists of list of conditions, a list of requests to apply if all the conditions are true, and a list of requests to apply if any condition is false.
+
+#### Options
+
+- hex -- print out keys and values as hex encoded string
+
+- i -- input transaction with interactive mode
```
Comment (xiang90) [inline etcdctlv3/README.md]: we probably need to make this (--interactive, -i).

Now it behavior incorrect
```
./etcdctlv3 txn -i
Error: unknown shorthand flag: 'i' in -i

OPTIONS:
      --hex[=false]	print out key-values as hex encoded strings
      --i[=false]	input transaction in interactive mode
```

File: etcdctlv3/command/txn_command.go [hunk 9]
```diff
@@ -41,198 +51,178 @@ func txnCommandFunc(cmd *cobra.Command, args []string) {
...
+
 	reader := bufio.NewReader(os.Stdin)
 
-	next := compareState
-	txn := &pb.TxnRequest{}
-	for next != nil {
-		next = next(txn, reader)
-	}
+	txn := clientv3.NewKV(mustClientFromCmd(cmd)).Txn(context.Background())
+	fmt.Println("entry comparison[key target expected_result compare_value] (end with empty line):")
```
Comment (xiang90) [inline etcdctlv3/command/txn_command.go]: update this hint to match current format?

File: etcdctlv3/command/txn_command.go [hunk 10]
```diff
@@ -41,198 +51,178 @@ func txnCommandFunc(cmd *cobra.Command, args []string) {
...
+	txn.If(readCompares(reader)...)
+	fmt.Println("entry success request[method key value(end_range)] (end with empty line):")
```
Comment (xiang90) [inline etcdctlv3/command/txn_command.go]: update this hint to match current format?

File: etcdctlv3/command/txn_command.go [hunk 11]
```diff
@@ -41,198 +51,178 @@ func txnCommandFunc(cmd *cobra.Command, args []string) {
...
+	txn.Then(readOps(reader)...)
+	fmt.Println("entry failure request[method key value(end_range)] (end with empty line):")
```
Comment (xiang90) [inline etcdctlv3/command/txn_command.go]: update this hint to match current format?

File: etcdctlv3/command/txn_command.go (hunk 9 above)
Comment (xiang90) [inline etcdctlv3/command/txn_command.go]: or simply say compares:

File: etcdctlv3/README.md [hunk 12]
```diff
@@ -109,6 +109,64 @@ OK
...
+
+- Additional error string if TXN failed. Exit code is non-zero.
+
+TODO: probably json and binary encoded proto
+
+#### Examples
+
+``` bash
+./etcdctl txn -i
+key1 greater mod 0
```
Comment (xiang90) [inline etcdctlv3/README.md]:
```
entry comparison[key target expected_result compare_value] (end with empty line):
key1 greater mod 0
Error:  invalid txn compare request: key1 greater mod 0
```

File: etcdctlv3/command/txn_command.go [hunk 13]
```diff
@@ -41,198 +51,178 @@ func txnCommandFunc(cmd *cobra.Command, args []string) {
...
-					RangeEnd: []byte(parts[2]),
-				}}
-		} else {
-			ru.Request = &pb.RequestUnion_RequestDeleteRange{
-				RequestDeleteRange: &pb.DeleteRangeRequest{
-					Key: key,
-				}}
-		}
+	cmpType := ""
+	switch parts[2] {
```
Comment (xiang90) [inline etcdctlv3/command/txn_command.go]: this should be part[1]?

bnf is <key> <CMPOP> <(CMPCREATE|CMPMOD|CMPVAL|CMPVER)>

File: etcdctlv3/command/txn_command.go [hunk 14]
```diff
@@ -41,198 +51,178 @@ func txnCommandFunc(cmd *cobra.Command, args []string) {
...
-	c := &pb.Compare{}
-	c.Key = []byte(parts[0])
+	var (
+		v   int64
+		err error
+		cmp clientv3.Cmp
+	)
+
+	key := parts[0]
 	switch parts[1] {
```
Comment (xiang90) [inline etcdctlv3/command/txn_command.go]: this should be part[2] according to bnf?

File: etcdctlv3/README.md (hunk 12 above)
Comment (xiang90) [inline etcdctlv3/README.md]: i think something like mod(key1) > 0 is more readable.

Comment (xiang90) [inline etcdctlv3/README.md]: tried
```
./etcdctlv3 txn --i
entry comparison[key target expected_result compare_value] (end with empty line):
foo mod greater 0

entry success request[method key value(end_range)] (end with empty line):
put foo bar

entry failure request[method key value(end_range)] (end with empty line):
put bar foo

Error:  rpc error: code = 3 desc = "etcdserver: key is not provided"
```

Seem like we missed the key when construct the cmp request for some reason.

File: etcdctlv3/README.md (hunk 7 above)
Comment (heyitsanthony) [inline etcdctlv3/README.md]: tagging block with `ebnf`

File: etcdctlv3/README.md (hunk 8 above)
Comment (heyitsanthony) [inline etcdctlv3/README.md]: OK. Throwing in a patch for watcher too.

File: etcdctlv3/command/txn_command.go (hunk 9 above)
Comment (heyitsanthony) [inline etcdctlv3/command/txn_command.go]: using `compares:`

File: etcdctlv3/command/txn_command.go (hunk 11 above)
Comment (heyitsanthony) [inline etcdctlv3/command/txn_command.go]: using `success requests (get, put, delete):`; wordy hint would be cluttered and probably wouldn't be any more helpful

Comment (xiang90) [inline etcdctlv3/command/txn_command.go]: agree

File: etcdctlv3/README.md (hunk 12 above)
Comment (heyitsanthony) [inline etcdctlv3/README.md]: ok, will change the format to `tgt(key) op val`

**Comment by heyitsanthony** [comment]
all fixed ptal /cc @xiang90

**Comment by xiang90** [comment]
LGTM

**Comment by heyitsanthony** [comment]
squashed, merging on greenlight

**Pull request merged**